File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/utils/UserMdcFilter.java
package uk.gov.justice.digital.hmpps.riskprofiler.utils;

import lombok.extern.slf4j.Slf4j;
import org.slf4j.MDC;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;
import uk.gov.justice.digital.hmpps.riskprofiler.security.UserSecurityUtils;

import javax.servlet.*;
import java.io.IOException;

import static uk.gov.justice.digital.hmpps.riskprofiler.utils.MdcUtility.USER_ID_HEADER;

@Slf4j
@Component
@Order(1)
public class UserMdcFilter implements Filter {

    private final UserSecurityUtils userSecurityUtils;

    @Autowired
    public UserMdcFilter(UserSecurityUtils userSecurityUtils) {
        this.userSecurityUtils = userSecurityUtils;
    }

    @Override
    public void init(FilterConfig filterConfig) {
        // Initialise - no functionality
    }

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
            throws IOException, ServletException {

        String currentUsername = userSecurityUtils.getCurrentUsername();

        try {
            if (currentUsername != null) {
                MDC.put(USER_ID_HEADER, currentUsername);
            }
            chain.doFilter(request, response);
        } finally {
            if (currentUsername != null) {
                MDC.remove(USER_ID_HEADER);
            }
        }
    }

    @Override
    public void destroy() {
        // Destroy - no functionality
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/utils/UserMdcFilter.kt
package uk.gov.justice.digital.hmpps.riskprofiler.utils

import jakarta.servlet.Filter
import jakarta.servlet.FilterChain
import jakarta.servlet.FilterConfig
import jakarta.servlet.ServletException
import jakarta.servlet.ServletRequest
import jakarta.servlet.ServletResponse
import org.slf4j.MDC
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component
import uk.gov.justice.digital.hmpps.riskprofiler.security.UserSecurityUtils
import java.io.IOException

@Component
@Order(1)
class UserMdcFilter @Autowired constructor(private val userSecurityUtils: UserSecurityUtils) : Filter {
  override fun init(filterConfig: FilterConfig) {
    // Initialise - no functionality
  }

  @Throws(IOException::class, ServletException::class)
  override fun doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain) {
    val currentUsername = userSecurityUtils.currentUsername
    try {
      if (currentUsername != null) {
        MDC.put(MdcUtility.USER_ID_HEADER, currentUsername)
      }
      chain.doFilter(request, response)
    } finally {
      if (currentUsername != null) {
        MDC.remove(MdcUtility.USER_ID_HEADER)
      }
    }
  }

  override fun destroy() {
    // Destroy - no functionality
  }
}
